third_party/amd/lib/TritonAMDGPUTransforms/UpdateAsyncWaitCount.cpp

```diff
@@ -6,130 +6,126 @@
 #define GEN_PASS_CLASSES
 #include "TritonAMDGPUTransforms/Passes.h"
 
+// This pass updates the waitCount of `AsyncWait` Ops to represent the number of
+// inflight async load operation between the async_wait and the definition of
+// the async_tokens. This value can then be used to wait only on the dependent
+// async loads before accessing the data allowing loads issues after to complete
+// in the future. This also means we should never overestimate the value to
+// ensure correctness wherease underestimating only affects performance.
+// So for each async_wait we need to compute the minimum across all async_token
+// operands.
+// For each token the minimum number of async transaction along it's def chain
+// is deduced. Note that a token can have multiple producers, e.g. if it's loop
+// carried (prologue and loop body). Therefore all paths to all producers of the
+// async_token have to be analyzed.
+// Note that we do not exit early if we encounter another async_wait along the
+// def chain because the pipeliner will merge redundant waits for us already
+
 using namespace mlir;
 namespace tt = triton;
 namespace ttg = triton::gpu;
 
-// LLVM cannot infer the dependency between direct to lds (async) loads and the
-// local reads between warps in a workgroup. As a workaround we update the
-// waitcnt to represent the number of hardware instructions we are interleaving
-// with. This allows us to manually emit the waitcnt during lowering.
-struct UpdateAsyncWaitCount : public OpRewritePattern<ttg::AsyncWaitOp> {
-  UpdateAsyncWaitCount(MLIRContext *ctx) : OpRewritePattern(ctx) {}
-
-  LogicalResult matchAndRewrite(ttg::AsyncWaitOp waitOp,
-                                PatternRewriter &rewriter) const override {
-    int waitCnt = std::numeric_limits<int>::max();
-
-    // AsyncWait can await multiple tokens so we get the minimum from all
-    // tokens
-    for (auto token : waitOp.getOperands()) {
-      // Traverse def chain from waitOp to the producer of the token and count
-      // the minumum number of vmcnt instructions
-      auto tokenWaitCnt =
-          findMinPathCountInDefChain(token, waitOp, [this](Operation *op) {
-            return getNumberOfLoadInstructions(op);
-          });
-      waitCnt = std::min(waitCnt, tokenWaitCnt);
-    }
-
-    if (waitCnt == std::numeric_limits<int>::max() ||
-        waitOp.getNum() == waitCnt)
-      return failure();
+// Overload to get the number of loads from direct to lds memory layouts
```

Comment:
"Returns the number of individual async load memory transactions when copy data from the given |srcTy| in global memory to the given |dstTy| in shared memory."